diffkemp/simpll/DifferentialFunctionComparator.cpp

```diff
@@ -940,26 +946,46 @@ int DifferentialFunctionComparator::cmpBasicBlocks(
             // it will be used in an inverse condition. Hence, we skip it here
             // and mark that it may be inverse-matching the condition that
             // has been originally mapped to the operand of the not operation.
-            if (isLogicalNot(&*InstL) || isLogicalNot(&*InstR)) {
+            if (config.Patterns.InverseConditions
+                && (isLogicalNot(&*InstL) || isLogicalNot(&*InstR))) {
+                // Note: This does not handle cases with a lot of `not`:
+                // %1 = cmp true, false | %1 = cmp true, false
+                //                      | %2 = not %1 (%2 = xor %1, true)
+                //                      | %3 = not %2 (%3 = xor %2, true)
+                //                      | %4 = not %3 (%4 = xor %3, true)
+                // br %1, l %T, l %F    | br %4, l %F, l %T
                 sn_mapL.erase(&*InstL);
                 sn_mapR.erase(&*InstR);
 
                 std::pair<const Value *, const Value *> matchingPair;
+                std::pair<const Value *, const Value *> oldMatchingPair;
                 if (isLogicalNot(&*InstL)) {
                     matchingPair = {&*InstL,
                                     getMappedValue(InstL->getOperand(0), true)};
+                    oldMatchingPair = {
+                            InstL->getOperand(0),
+                            getMappedValue(InstL->getOperand(0), true)};
                     replacedInstructions.emplace(&*InstL, InstL->getOperand(0));
                     InstL++;
                 } else {
                     matchingPair = {getMappedValue(InstR->getOperand(0), false),
                                     &*InstR};
+                    oldMatchingPair = {
+                            getMappedValue(InstR->getOperand(0), false),
+                            InstR->getOperand(0)};
                     replacedInstructions.emplace(&*InstR, InstR->getOperand(0));
                     InstR++;
                 }
 
                 // If the conditions are already inverse, remove them from the
                 // list. Otherwise, add them.
-                size_t erased = inverseConditions.erase(matchingPair);
+                size_t erased = inverseConditions.erase(oldMatchingPair);
+                LOG_VERBOSE_EXTRA((erased ? "Removed" : " Added")
+                                  << " inverse condition\n");
+                LOG_INDENT();
+                LOG_VERBOSE_EXTRA("L: " << (*(matchingPair.first)) << "\n");
+                LOG_VERBOSE_EXTRA("R: " << (*(matchingPair.second)) << "\n");
+                LOG_UNINDENT();
```

Comment:
This visually decouples the two related operations of updating `inverseConditions`. Better move it below the `insert` line?

Reply:
Thanks, I edited it and I think it is better.